src/SmoFacade/Listener.cs

```diff
@@ -130,19 +130,70 @@ private static string AgListenerName(string dataSource)
-      else if(parts.Length == 2)
-        // NamedInstances: chop instance name, resolve DNS, slap instance name back on!
-        connBuilder.DataSource = $"{Dns.GetHostEntry(parts[0]).HostName}\\{parts[1]}";
-      else
-        throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS");
-
-      return new Server(connBuilder.ToString(), credentialName);
+      try 
+      {
+        connBuilder.DataSource = ResolveDnsHostNameForInstance(agInstanceName, connBuilder.DataSource);
+        return new Server(connBuilder.ToString(), credentialName);
+      }
+      catch (Exception e)
+      {
+        throw new ArgumentException($"agInstanceName param {agInstanceName} cannot be resolved by DNS", e);
+      }
+    }
+
+    /// <summary>
+    ///   Resolves 'agReplicaInstanceName' to a FQDN
+    ///   However on Unix OS, when 'val' in 'Dns.GetHostEntry(val)' is not a complete domain (i.e is just "abc", instead of "abc.def.com"), it fails intermittently
+    ///   Therefore, if dns lookup on just the instance name fails, we retry after appending the domain fragments from the listener to the instance name
+    /// </summary>
+    /// <param name="agReplicaInstanceName"> The name for an instance within the AG (for which we are trying to get the FQDN)</param>
+    /// <param name="agListenerDomain"> The complete domain for the AG listener</param>
+    private static string ResolveDnsHostNameForInstance(string agReplicaInstanceName, string agListenerDomain)
+    {
+      // Sometimes instances and listeners have ports(or "named instances")
+      // Therefore, we strip them off before calling DNS.GetHostEntry() and then add them back to the result 
+      var (listenerDomain, listenerPort) = SplitDomainAndPort(agListenerDomain);
+      var (instanceName, instancePort) = SplitDomainAndPort(agReplicaInstanceName);
+      // preference is to add back 'instancePort' over 'listenerPort' (in almost all cases they should be identical)
+      var port = instancePort ?? listenerPort;
+
+      try
+      {
+        return $"{Dns.GetHostEntry(instanceName).HostName}{port}";
+      }
+      catch (System.Net.Sockets.SocketException)
+      {
+        // Re-try by appending the domain fragments from listener to the instance name 
+        // However, we don't need the listener's "host name" (first fragment), so we need to strip that off
+        // eg: if listener is "abc.def.ghi" we want to append only ".def.ghi" to the instance name
+        var listenerDomainFragments = listenerDomain.Split('.');
+        listenerDomainFragments[0] = null;
+        var instanceDomain = $"{instanceName}{string.Join(".", listenerDomainFragments)}";
+
+        return $"{Dns.GetHostEntry(instanceDomain).HostName}{port}";
+      }
     }
+
+    // This function handles named instances ("<domain>\<named instance>") in the same way as ports
+    private static (string, string) SplitDomainAndPort(string domainAndPort)
+    {

```

Comment:
@alexirion10 working on Unit tests for this - was wondering if we should have some for `ResolveDnsHostNameForInstance()` too to check it retries correctly?

Or is just checking this one fine?

Reply:
the more tests the better